src/core/lib/Bombe.mjs

```diff
@@ -0,0 +1,472 @@
+        graphs.sort((a, b) => {
+            let result = b[0] - a[0];
+            if (result === 0) {
+                result = b[1] - a[1];
+            }
+            return result;
+        });
+        this.nLoops = graphs[0][0];
+        return [graphs[0][2], graphs[0][4]];
+    }
+
+    /**
+     * Implement Welchman's diagonal board: If A steckers to B, that implies B steckers to A, and
+     * so forth. This function just gets the paired wire.
+     * @param {number[2]} i - Bombe state wire
+     * @returns {number[2]}
+     */
+    getDiagonal(i) {
+        return [i[1], i[0]];
+    }
+
+    /**
+     * Bombe electrical simulation. Energise a wire. For all connected wires (both via the diagonal
+     * board and via the scramblers), energise them too, recursively.
+     * @param {number[2]} i - Bombe state wire
+     */
+    energise(i) {
+        const idx = 26*i[0] + i[1];
+        if (this.wires[idx]) {
+            return;
+        }
+        this.energiseCount ++;
+        this.wires[idx] = true;
+        this.energise(this.getDiagonal(i));
+
+        for (const scrambler of this.scramblers[i[0]]) {
+            const out = scrambler.transform(i[1]);
+            const other = scrambler.getOtherEnd(i[0]);
+            this.energise([other, out]);
+        }
+    }
+
+    /**
+     * Having set up the Bombe, do the actual attack run. This tries every possible rotor setting
+     * and attempts to logically invalidate them. If it can't, it's added to the list of candidate
+     * solutions.
+     * @returns {string[][2]} - list of pairs of candidate rotor setting, and calculated stecker pair
+     */
+    run() {
+        let stops = 0;
+        const result = [];
+        // For each possible rotor setting
+        const nChecks = Math.pow(26, this.baseRotors.length);
+        for (let i=1; i<=nChecks; i++) {
+            this.wires.fill(false);
+            // Energise the test input, follow the current through each scrambler
+            // (and the diagonal board)
+            this.energiseCount = 0;
+            this.energise(this.testInput);
+            // Count the energised outputs
+            let count = 0;
+            for (let j=26*this.testRegister; j<26*(1+this.testRegister); j++) {
+                if (this.wires[j]) {
+                    count++;
+                }
+            }
+            // If it's not all of them, we have a stop
+            if (count < 26) {
+                stops += 1;
+                let stecker;
+                // The Bombe tells us one stecker pair as well. The input wire and test register we
+                // started with are hypothesised to be a stecker pair.
+                if (count === 25) {
+                    // Our steckering hypothesis is wrong. Correct value is the un-energised wire.
+                    for (let j=0; j<26; j++) {
+                        if (!this.wires[26*this.testRegister + j]) {
```

Comment:
Considering this is a nested loop, performance is very important here.

It's possible you could use something along these lines instead.

``` javascript
const unEnergisedWire = this.wires.indexOf(false, 26*this.testRegister) - 26*this.testRegister;
```

The problem here would be if none of the wires are energised, in which case `indexOf` will keep looking through the `wires` array beyond 26 places. There [isn't a way](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/indexOf) to limit `indexOf`.

Reply:
Of course, if we haven't initialised the `wires` array, we might have to search for `undefined` instead.